src/tokenserver/auth/browserid.rs

```diff
@@ -0,0 +1,469 @@
+use core::time::Duration;
+
+use async_trait::async_trait;
+use reqwest::{blocking::Client as ReqwestClient, StatusCode};
+use serde::{de::Deserializer, Deserialize, Serialize};
+
+use super::VerifyToken;
+use crate::tokenserver::{
+    error::{ErrorLocation, TokenserverError},
+    settings::Settings,
+};
+
+/// The information extracted from a valid BrowserID assertion.
+#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
+pub struct VerifyOutput {
+    pub device_id: Option<String>,
+    pub email: String,
+    pub generation: Option<i64>,
+    pub keys_changed_at: Option<i64>,
+}
+
+/// The verifier used to verify BrowserID assertions.
+#[derive(Clone)]
+pub struct RemoteVerifier {
+    audience: String,
+    issuer: String,
+    fxa_verifier_url: String,
+    request_client: ReqwestClient,
+}
+
+impl From<&Settings> for RemoteVerifier {
+    fn from(settings: &Settings) -> Self {
+        Self {
+            audience: settings.fxa_browserid_audience.clone(),
+            issuer: settings.fxa_browserid_issuer.clone(),
+            request_client: ReqwestClient::builder()
+                .timeout(Duration::new(settings.fxa_browserid_request_timeout, 0))
+                .build()
+                .expect("Couldn't initialize BrowserID verifier request client"),
+            fxa_verifier_url: settings.fxa_browserid_server_url.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl VerifyToken for RemoteVerifier {
+    type Output = VerifyOutput;
+
+    /// Verifies a BrowserID assertion. Returns `VerifyOutput` for valid assertions and a
+    /// `TokenserverError` for invalid assertions.
+    async fn verify(&self, assertion: String) -> Result<VerifyOutput, TokenserverError> {
+        let response = self
+            .request_client
+            .post(&self.fxa_verifier_url)
+            .json(&VerifyRequest {
+                assertion,
+                audience: self.audience.clone(),
+                trusted_issuers: [self.issuer.clone()],
+            })
+            .send()
+            .map_err(|e| {
+                if e.is_connect() {
+                    // If we are unable to reach the FxA server, report a 503 to the client
+                    TokenserverError::resource_unavailable()
+                } else {
+                    // If any other error occurs during the request, report a 401 to the client
+                    TokenserverError::invalid_credentials("Unauthorized")
+                }
+            })?;
+
+        // If FxA responds with an HTTP status other than 200, report a 503 to the client
+        if response.status() != StatusCode::OK {
+            return Err(TokenserverError::resource_unavailable());
+        }
+
+        // If FxA responds with an invalid response body, report a 503 to the client
+        let response_body = response
+            .json::<VerifyResponse>()
```

Comment:
Due to serde/Rust's type system, there is some deviation between how the old and new Tokenserver handles a faulty FxA response. For certain situations (e.g. a `fxa-generation` claim that isn't an integer), the old Tokenserver returns a 401, but for others (e.g. FxA responds with malformed JSON), it returns a 503. I would have had to write a pretty lengthy custom `Deserializer` to handle these cases the same way, and I figured it would be better just to handle all of them as 503 errors anyway, since they _would_ be errors internal to Sync.